### ISSUE TITLE
Sensor query for user power limit always returns 0 to HTMGT

### ISSUE THREAD

**Issue by cdeadmin** [openbmc/openbmc#3012]
<cde:info> SW421108 </cde:info>

**Comment by amboar** [comment]
Can you please provide a description of how to reproduce this issue, and what versions are affected?

**Comment by ngorugan** [comment]
@amboar Below given info  is provided by @cjcain in CQ.

All queries of the user power limit from HTMGT always return 0.  The user power limit status
does get returned correctly.

When I set a user power limit (via curl commands), the BMC appears to be processing those
command and does return the correct value when queried via curl. OCC is also informed of the
new limit:
```
$ curl -c cjar -b cjar -k -H "Content-Type: application/json" https://$BMCIP/xyz/openbmc_project/control/host0/power_cap
{
 "data": {
    "PowerCap": 0,
    "PowerCapEnable": 0
 },
 "message": "200 OK",
 "status": "ok"
}

$ curl -c cjar -b cjar -k -H "Content-Type: application/json"  -X PUT -d '{"data":1500}' https://$BMCIP/xyz/openbmc_project/control/host0/power_cap/attr/PowerCap
{
 "data": null,
 "message": "200 OK",
 "status": "ok"
}

$ curl -c cjar -b cjar -k -H "Content-Type: application/json"https://$BMCIP/xyz/openbmc_project/control/host0/power_capp
{
 "data": {
    "PowerCap": 1500,
    "PowerCapEnable": 0
 },
 "message": "200 OK",
 "status": "ok"
}

$ curl -c cjar -b cjar -k -H "Content-Type: application/json"  -X PUT -d '{"data":1}' https://$BMCIP/xyz/openbmc_project/control/host0/power_cap/attr/PowerCapEnable
{
 "data": null,
 "message": "200 OK",
 "status": "ok"
}

$ curl -c cjar -b cjar -k -H "Content-Type: application/json"https://$BMCIP/xyz/openbmc_project/control/host0/power_cape
{
 "data": {
    "PowerCap": 1500,
    "PowerCapEnable": 1
 },
 "message": "200 OK",
 "status": "ok"
}
```
```
$ ./occtoolp9 -p
Sending 0x00 command to OCC0 (via opal-prd)...
 Data Length: 0x034D
       Status: 0xC3 ( Master FIRMaster ObsReady ActReady)
    ExtStatus: 0x10 ( QckPwrDrop )
         OCCs: 0x03
      CfgNeed: 0x00
        State: 0x03 (ACTIVE)
...
       Sensor: CAPS - format:3 / 1 sensors (15 bytes/sensor)
         Current Power Cap:   1350 Watts
             Current Power:    328 Watts (output power)
               N Power Cap:   1700 Watts (cap without redundant power)
             Max Power Cap:   3050 Watts
        Hard Min Power Cap:   1700 Watts
        Soft Min Power Cap:    500 Watts
            User Power Cap:   1350 Watts
        User Power Limit Source: 1
...

```
So the above is all good.  The problem occurs when the OCC gets reset, or the system gets re-ipled.
After the OCC reset happens or when the system reboots, HTMGT will re-query the BMC for the
user power limit and status.  The response that HTMGT gets in both of these cases is 0 for the watt limit.  The status appears to be correct.

Console trace with power limit disabled (looks fine)
```
104.68306|htmgt|OCC1 Poll change: Status:0310 Occs:02 Cfg:00 State:01 Error:000000/00000000
104.68307|htmgt|OCC1 Poll change: Status:0310 Occs:02 Cfg:00 State:01 Error:000000/00000000
104.92105|htmgt|SENSOR::getUserPowerLimit returned 0W, active: N
104.92106|htmgt|sendOccUserPowerCap: Sending power cap 0W to OCC 0
104.92107|htmgt|Send OCC0 cmd=0x22 (SET_POWER_CAP), length=0x0002
104.92108|htmgt|OCC0 Command: SET_POWER_CAP (0x0008 bytes)
104.92109|htmgt|0000:  1E220002 00000042 00920000 000000C6
105.00919|htmgt|waitForOccRsp: OCC0 rsp (rsp length=7)
105.00920|htmgt|OCC0 Response: (0x0007 bytes)
105.00920|htmgt|0000:  1E220000 00004000 00000000 00000000
105.01447|htmgt|OCC0 rsp status=0x00, length=0x0000

Console trace with power limit enabled and set to 1500W: (BAD)
59.61320|htmgt|OCC1 Poll change: Status:0310 Occs:02 Cfg:00 State:01 Error:000000/00000000
59.61322|htmgt|OCC1 Poll change: Status:0310 Occs:02 Cfg:00 State:01 Error:000000/00000000
59.80825|htmgt|SENSOR::getUserPowerLimit returned 0W, active: Y
59.80827|htmgt|getMaxPowerCap: maximum power cap = 3050W (redundant PS bulk power limit)
59.80827|htmgt|sendOccUserPowerCap: Sending power cap 0W to OCC 0
59.80829|htmgt|Send OCC0 cmd=0x22 (SET_POWER_CAP), length=0x0002
59.80830|htmgt|OCC0 Command: SET_POWER_CAP (0x0008 bytes)
59.80830|htmgt|0000:  1E220002 00000042 00920000 000000C6
59.92222|htmgt|waitForOccRsp: OCC0 rsp (rsp length=7)
59.92223|htmgt|OCC0 Response: (0x0007 bytes)
59.92224|htmgt|0000:  1E220000 00004000 00000000 00000000
59.91651|htmgt|OCC0 rsp status=0x00, length=0x0000
```
I disabled and re-enabled the power limit via curl, and the OCC again reflected the limit.
I issued an PM Complex Reset which causes HTMGT to re-query the user power limit, but it again
got a value of 0.  Here is the trace during that query:
```
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: HTMGT:OCC1 Poll change: Status:0310 Occs:02 Cfg:00 State:01 Error:000000/00000000
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: IPMI:rt: calling sync 2c:3  len=3
Mar 14 15:18:49 ws003p1 opal-prd: IPMI: sending 3 bytes (netfn 0x2c, cmd 0x03)
Mar 14 15:18:49 ws003p1 opal-prd: IPMI: received 3 bytes
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: IPMI:rt: calling sync 4:2d  len=1
Mar 14 15:18:49 ws003p1 opal-prd: IPMI: sending 1 bytes (netfn 0x04, cmd 0x2d)
Mar 14 15:18:49 ws003p1 opal-prd: IPMI: received 1 bytes
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: IPMI:Derating factor = 0
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: IPMI:Power limit = 0x0
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: IPMI:Power limit is active
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: HTMGT:SENSOR::getUserPowerLimit returned 0W, active: Y
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: HTMGT:getMaxPowerCap: maximum power cap = 3050W (redundant PS bulk power limit)
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: HTMGT:sendOccUserPowerCap: Sending power cap 0W to OCC 0
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: HTMGT:Send OCC0 cmd=0x22 (SET_POWER_CAP), length=0x0002
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: HTMGT:cmd data (up to 16 bytes)
Mar 14 15:18:49 ws003p1 opal-prd: HBRT: HTMGT:~[0x0000] 0000                                   *..              *
```

System I was testing on was wsbmxxx.xxx.xxx
```
# cat /etc/os-release
ID="openbmc-phosphor"
NAME="Phosphor OpenBMC (Phosphor OpenBMC Project Reference Distro)"
VERSION="ibm-v2.0"
VERSION_ID="ibm-v2.0-0-r41-0-gd0c319e"
PRETTY_NAME="Phosphor OpenBMC (Phosphor OpenBMC Project Reference Distro) ibm-v2.0"
BUILD_ID="ibm-v2.0-0-r41"

# cat /var/lib/phosphor-software-manager/pnor/ro/VERSION
open-power-witherspoon-v1.21-55-g309da0f-dirty
        buildroot-2017.11.2-8-g4b6188e
        skiboot-v5.10-45-g84186ef0944c
        hostboot-f5d2c87
        linux-4.15.6-openpower1-pc9a8340
        petitboot-v1.7.1-p78f3192
        machine-xml-c10638f-pc56f45a
        occ-2fe8f2c
        hostboot-binaries-290a4f0
        capp-ucode-p9-dd2-v3
        sbe-75ddac2
```

**Comment by amboar** [comment]
@ngorugan thanks for the detail, that makes it much clearer. However, is it possible to hit this issue on the github master builds? The VERSION_ID above shows that this was reported against an internal IBM build.

**Comment by rfrandse** [comment]
<https://gerrit.openbmc-project.xyz/10499> Fix Sensor query for derating factor returns 0
Resolves  #3012 Sensor query for user power limit always returns 0 to HTMGT
